### PR TITLE
Feat/templating add self-running templates

### PR DESCRIPTION
**Pull request by dwertheimer**
Imagine an xcallback that gives users the ability to "run" any template to perform an action...
Here's a simple example which will open today's note, add some text under a certain heading. All passed through callback and hitting a user-created Template. No plugin code required.

1. Create a template like this:
```
---
title: cdjournal
openNoteTitle:  <TODAY>
writeUnderHeading: "📝 Notes & Goals"
location: prepend
---
**<%- date.timestamp() %>**
<%- journalEntry %>
```
2. Run this link
```
noteplan://x-callback-url/runPlugin?pluginID=np.Templating&command=templateRunner&arg0=cdjournal&arg1=true&arg2=journalEntry%3dYes%20IT%20WORKED
```
3. It should open today's daily note and you should see text under your Notes & Goals heading

Known issue: If the heading doesn't exist, it's not working. I think this is an API bug...still chasing it.

### CODE REVIEW

File: np.Templating/plugin.json [hunk 1]
```diff
@@ -98,6 +98,22 @@
...
+    {
+      "name": "np:tft",
+      "alias": [
+        "Template file by title"
+      ],
+      "description": "Run a template file by title/preset",
+      "jsFunction": "templateFileByTitle"
+    },
+    {
+      "name": "templateRunner",
```
Comment (mikeerickson) [inline np.Templating/plugin.json:111]: This goes against the naming conventions used for all commands.  You have it right in the previous command.

e.g.

"name: "np:tr"

File: np.Templating/plugin.json [hunk 2]
```diff
@@ -98,6 +98,22 @@
...
+      "alias": [
+        "np:tr"
```
Comment (mikeerickson) [inline np.Templating/plugin.json:113]: you could use "templateRunner" here an alias

File: np.Templating/src/Templating.js [hunk 3]
```diff
@@ -273,6 +273,116 @@ export async function templateNew(): Promise<void> {
...
+      let note
+      if (isTodayNote) {
+        await Editor.openNoteByDate(new Date(), false)
+        await writeNoteContents(Editor.note, renderedTemplate, writeUnderHeading, shouldAppend)
+      } else {
+        if (noteTitle.length) {
+          const notes = await DataStore.projectNoteByTitle(noteTitle)
+          if (!notes || notes.length == 0 || (notes && notes.length > 1)) {
+            const msg = `${!notes || notes.length == 0 ? 'no' : notes.length > 1 ? 'more than one' : ''}`
+            await CommandBar.prompt(`"${noteTitle}" matches ${msg} note(s) title. the title needs to be distinct to ensure correct note is written to.`, helpInfo('Presets'))
```
Comment (mikeerickson) [inline np.Templating/src/Templating.js:346]: Rephase to use `unique` instead of `distinct` more "human", less "techy" :-)

**Review by mikeerickson — CHANGES_REQUESTED** [review]
All in all, seems pretty clean and straight forward.  I will work on getting this implemented after the 1.2 release though so that I can have proper testing time (I will be releasing 1.2.0-beta.01 today, had to wait until latest NotePlan Alpha to make sure the `invokePluginCommandByName` bug was fixed (which it was)

File: np.Templating/plugin.json (hunk 1 above)
Comment (dwertheimer) [reply np.Templating/plugin.json:111]: Thx for the review @mikeerickson. Great comments. w/r/t the commands, I did this on purpose though. When we are finished testing, I plan to hide these commands from the UX, so no user will ever see them as `np:*` commands. But I needed a reasonable command name for users to use in the x-callback: `noteplan://x-callback-url/runPlugin?pluginID=np.Templating&command=templateRunner`, and I'm a bit afraid of colons in the xcallback context. Does that make sense or do you still want me to change it? `noteplan://x-callback-url/runPlugin?pluginID=np.Templating&command=np:tr`

File: np.Templating/src/Templating.js (hunk 3 above)
Comment (dwertheimer) [reply np.Templating/src/Templating.js:346]: agreed. done.

File: np.Templating/src/Templating.js [hunk 4]
```diff
@@ -273,6 +273,116 @@ export async function templateNew(): Promise<void> {
...
+          } else {
+            note = notes[0] || null
+            if (!note) {
+              await CommandBar.prompt(`Unable to locate note matching: "${noteTitle}"`, helpInfo('Presets'))
+            } else {
+              await writeNoteContents(note, renderedTemplate, writeUnderHeading, shouldAppend, shouldOpenInEditor)
+            }
+          }
+        } else {
+          await CommandBar.prompt(`openNoteTitle or writeNotetitle is required`, helpInfo('Presets'))
```
[the remark this answers is not in the document]
Comment (dwertheimer) [reply np.Templating/src/Templating.js:356]: done

File: np.Templating/src/Templating.js [hunk 5]
```diff
@@ -273,6 +273,116 @@ export async function templateNew(): Promise<void> {
...
+/**
+ * Write out the contents to either Today's Calendar note or the Note which was opened
+ * @author @dwertheimer
+ * @param {*} note
+ * @param {*} renderedTemplate
+ * @param {*} writeUnderHeading
+ * @param {*} shouldAppend
+ * @param {*} shouldOpenInEditor
+ */
+async function writeNoteContents(note, renderedTemplate, writeUnderHeading, shouldAppend, shouldOpenInEditor): Promise<void> {
```
[the remark this answers is not in the document]
Comment (dwertheimer) [reply np.Templating/src/Templating.js:285]: Yes, that's fair. Was moving too fast in refactoring. Will do. BTW, this is 5 params. Would you prefer that I change the signature to collapse some of these into an options object?

File: np.Templating/CHANGELOG.md [hunk 6]
```diff
@@ -94,6 +94,10 @@ Also, it should be encouraged to rename "📋 Templates" to another name "📋 T
 - added default date/time prompt value when when using `np:mtn` (uses `timestampFormat` format in `np.Templating` settings)
 - updated README command reference
 
+## [1.1.0] - 2022-04-15 (dwertheimer)
```
[the remark this answers is not in the document]
Comment (dwertheimer) [reply np.Templating/CHANGELOG.md:97]: OK I will wait until you push again and merge. Pretty sure I pulled main before I started work on this.

File: np.Templating/src/Templating.js [hunk 7]
```diff
@@ -273,6 +273,116 @@ export async function templateNew(): Promise<void> {
...
+ * Note: ask CD what the reserved frontmatter fields should be and trap for them
+ * xcallback note: arg1 is template name, arg2 is whether to open in editor, arg3 is a list of vars to pass to template equals sign is %3d
+ *
+ */
+export async function templateFileByTitle(selectedTemplate?: string, openInEditor?: boolean = false, args?: string = ''): Promise<void> {
+  try {
+    let argObj = {}
+    args.split(',').forEach((arg) => (arg.split('=').length === 2 ? (argObj[arg.split('=')[0]] = arg.split('=')[1]) : null))
+    if (!selectedTemplate || selectedTemplate.length === 0) {
+      await CommandBar.prompt(`templateFileByTitle: no templateTitle was specified."`, helpInfo('Presets'))
```
[the remark this answers is not in the document]
Comment (dwertheimer) [reply np.Templating/src/Templating.js:317]: done

**Comment by dwertheimer** [comment]
Made those changes and pushed an updated version. No hurry @mikeerickson. Get to it whenever.

File: np.Templating/CHANGELOG.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikeerickson) [reply np.Templating/CHANGELOG.md:113]: I am curious where the "old" `CHANGELOG.md` content is coming from, it is surely current on the current repo?

https://github.com/NotePlan/plugins/blob/main/np.Templating/CHANGELOG.md

File: np.Templating/plugin.json (hunk 1 above)
Comment (mikeerickson) [reply np.Templating/plugin.json:118]: Ah gotcha, then nothing to see here :-)

**Comment by dwertheimer** [comment]
@mikeerickson Bummer to have these PRs open so long cuz I am super bad at merging and now I don't know how I would merge this.

**Comment by mikeerickson** [comment]
Implemented in next version, cherry-picking code as opposed to merging PR as too many merge conflicts with current np.Templating code, which is diverged from the current release in which this PR was branched.

**Pull request closed without merging**